api/controllers/export-data-2.js

```diff
@@ -0,0 +1,257 @@
+  const rows = result.rows;
+
+  const contactIds = [];
+  rows.forEach(row => {
+    let parent = row.doc.contact;
+    while(parent) {
+      if (parent._id) {
+        contactIds.push(parent._id);
+      }
+      parent = parent.parent;
+    }
+  });
+
+  if (!contactIds.length) {
+    return Promise.resolve(result);
+  }
+
+  return DB.allDocs({
+    keys: contactIds,
+    include_docs: true
+  }).then(results => {
+    rows.forEach(row => {
+      const contactId = row.doc.contact && row.doc.contact._id;
+      const contact = findContact(results.rows, contactId);
+      if (contact) {
+        row.doc.contact = contact.doc;
+      }
+      let parent = row.doc.contact;
+      while(parent) {
+        const parentId = parent.parent && parent.parent._id;
+        const found = findContact(results.rows, parentId);
+        if (found) {
+          parent.parent = found.doc;
+        }
+        parent = parent.parent;
+      }
+    });
+
+    return result;
+  });
+};
+
+
+/**
+ * Flattens a given object into an object where the keys are dot-notation
+ * paths to the flattened values:
+ * {
+ *   "foo": {
+ *     "bar": "smang"
+ *   }
+ * }
+ *
+ * becomes:
+ *
+ * {
+ *   "foo.bar": "smang"
+ * }
+ */
+// TODO: is there really not a utility that does this?
+const flatten = (fields, prepend=[]) =>
+  Object.keys(fields).reduce((acc, k) => {
+    const path = [...prepend, k];
+    if (typeof fields[k] === 'object' && fields[k]) {
+      return _.extend(acc, flatten(fields[k], path));
+    } else {
+      acc[path.join('.')] = fields[k] || '';
+      return acc;
+    }
+  }, {});
+
+// TODO: is there really not a utility that does this?
+const safeGet = (obj, fields) => {
+  let v = obj;
+  for (const f of fields) {
+    v = v[f];
+    if (!v) {
+      return undefined;
+    }
+  }
+  return v;
+};
+
+const CSV_MAPPER = {
+  reports: (filters) => {
+    const forms = (
+      filters &&
+      filters.forms &&
+      filters.forms.selected &&
+      _.pluck(filters.forms.selected, 'code'));
+
+    let p = Promise.resolve();
+    // If they have not selected any forms we need to get all available ones
+    if (!forms) {
+      p = p.then(() =>
+        DB.query('medic-client/reports_by_form', {
+          group: true
+        }).then(results => results.rows.map(r => r.key[0]))
+      );
+    } else {
+      p = p.then(() => forms);
+    }
+
+    // Determine the fields for each form we care about
+    return p.then(forms =>
+      Promise.all(forms.map(form =>
+        DB.query('medic-client/reports_by_form', {
+          key: [form],
+          limit: 1,
+          include_docs: true,
+          reduce: false
+        })
+        .then(results =>
+          results.rows[0] &&
+          results.rows[0].doc &&
+          results.rows[0].doc.fields
+        ))
+      ).then(allFields => {
+        const fieldColumns = _.union(
+          ...allFields.filter(f => f).map(f => Object.keys(flatten(f)))
+        ).sort();
```

Comment:
Same thoughts as last time, but done!

I ended up splitting the logic a bit too, because understanding the `.filter(f => f)` is reliant on seeing where `allFields` comes from. I also changed it to `_.identity` because we're already using underscore and it's a word people can google.